// @mariozechner/pi-coding-agent/dist/oauth/index.js
import { loginAnthropic, refreshAnthropicToken } from "./anthropic.js";
import { listOAuthProviders as listOAuthProvidersFromStorage, loadOAuthCredentials, removeOAuthCredentials, saveOAuthCredentials, } from "./storage.js";
// Re-export for convenience
export { listOAuthProvidersFromStorage as listOAuthProviders };
/**
 * Get list of OAuth providers
 */
export function getOAuthProviders() {
    return [
        {
            id: "anthropic",
            name: "Anthropic (Claude Pro/Max)",
            available: true,
        },
        {
            id: "github-copilot",
            name: "GitHub Copilot (coming soon)",
            available: false,
        },
    ];
}
/**
 * Login with OAuth provider
 */
export async function login(provider, onAuthUrl, onPromptCode) {
    switch (provider) {
        case "anthropic":
            await loginAnthropic(onAuthUrl, onPromptCode);
            break;
        case "github-copilot":
            throw new Error("GitHub Copilot OAuth is not yet implemented");
        default:
            throw new Error(`Unknown OAuth provider: ${provider}`);
    }
}
/**
 * Logout from OAuth provider
 */
export async function logout(provider) {
    removeOAuthCredentials(provider);
}
/**
 * Refresh OAuth token for provider
 */
export async function refreshToken(provider) {
    const credentials = loadOAuthCredentials(provider);
    if (!credentials) {
        throw new Error(`No OAuth credentials found for ${provider}`);
    }
    let newCredentials;
    switch (provider) {
        case "anthropic":
            newCredentials = await refreshAnthropicToken(credentials.refresh);
            break;
        case "github-copilot":
            throw new Error("GitHub Copilot OAuth is not yet implemented");
        default:
            throw new Error(`Unknown OAuth provider: ${provider}`);
    }
    // Save new credentials
    saveOAuthCredentials(provider, newCredentials);
    return newCredentials.access;
}
/**
 * Get OAuth token for provider (auto-refreshes if expired)
 */
export async function getOAuthToken(provider) {
    const credentials = loadOAuthCredentials(provider);
    if (!credentials) {
        return null;
    }
    // Check if token is expired (with 5 min buffer already applied)
    if (Date.now() >= credentials.expires) {
        // Token expired - refresh it
        try {
            return await refreshToken(provider);
        }
        catch (error) {
            console.error(`Failed to refresh OAuth token for ${provider}:`, error);
            // Remove invalid credentials
            removeOAuthCredentials(provider);
            return null;
        }
    }
    return credentials.access;
}
