// @mariozechner/pi-coding-agent/dist/settings-manager.js
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "fs";
import { dirname, join } from "path";
import { getAgentDir } from "./config.js";
export class SettingsManager {
    settingsPath;
    settings;
    constructor(baseDir) {
        const dir = baseDir || getAgentDir();
        this.settingsPath = join(dir, "settings.json");
        this.settings = this.load();
    }
    load() {
        if (!existsSync(this.settingsPath)) {
            return {};
        }
        try {
            const content = readFileSync(this.settingsPath, "utf-8");
            return JSON.parse(content);
        }
        catch (error) {
            console.error(`Warning: Could not read settings file: ${error}`);
            return {};
        }
    }
    save() {
        try {
            // Ensure directory exists
            const dir = dirname(this.settingsPath);
            if (!existsSync(dir)) {
                mkdirSync(dir, { recursive: true });
            }
            writeFileSync(this.settingsPath, JSON.stringify(this.settings, null, 2), "utf-8");
        }
        catch (error) {
            console.error(`Warning: Could not save settings file: ${error}`);
        }
    }
    getLastChangelogVersion() {
        return this.settings.lastChangelogVersion;
    }
    setLastChangelogVersion(version) {
        this.settings.lastChangelogVersion = version;
        this.save();
    }
    getDefaultProvider() {
        return this.settings.defaultProvider;
    }
    getDefaultModel() {
        return this.settings.defaultModel;
    }
    setDefaultProvider(provider) {
        this.settings.defaultProvider = provider;
        this.save();
    }
    setDefaultModel(modelId) {
        this.settings.defaultModel = modelId;
        this.save();
    }
    setDefaultModelAndProvider(provider, modelId) {
        this.settings.defaultProvider = provider;
        this.settings.defaultModel = modelId;
        this.save();
    }
    getQueueMode() {
        return this.settings.queueMode || "one-at-a-time";
    }
    setQueueMode(mode) {
        this.settings.queueMode = mode;
        this.save();
    }
    getTheme() {
        return this.settings.theme;
    }
    setTheme(theme) {
        this.settings.theme = theme;
        this.save();
    }
    getDefaultThinkingLevel() {
        return this.settings.defaultThinkingLevel;
    }
    setDefaultThinkingLevel(level) {
        this.settings.defaultThinkingLevel = level;
        this.save();
    }
    getCompactionEnabled() {
        return this.settings.compaction?.enabled ?? true;
    }
    setCompactionEnabled(enabled) {
        if (!this.settings.compaction) {
            this.settings.compaction = {};
        }
        this.settings.compaction.enabled = enabled;
        this.save();
    }
    getCompactionReserveTokens() {
        return this.settings.compaction?.reserveTokens ?? 16384;
    }
    getCompactionKeepRecentTokens() {
        return this.settings.compaction?.keepRecentTokens ?? 20000;
    }
    getCompactionSettings() {
        return {
            enabled: this.getCompactionEnabled(),
            reserveTokens: this.getCompactionReserveTokens(),
            keepRecentTokens: this.getCompactionKeepRecentTokens(),
        };
    }
    getHideThinkingBlock() {
        return this.settings.hideThinkingBlock ?? false;
    }
    setHideThinkingBlock(hide) {
        this.settings.hideThinkingBlock = hide;
        this.save();
    }
}
